### PR TITLE
Charactercls

### PR DESCRIPTION
**Pull request by mumme74**
This is also a part of pull request #74

### CODE REVIEW

File: peglib.h [hunk 1]
```diff
@@ -389,16 +390,23 @@ inline std::string resolve_escape_sequence(const char* s, size_t n) {
     while (i < n) {
         auto ch = s[i];
         if (ch == '\\') {
+            if (i+1 == n) { // if last char is '\' it should return '\'
```
Comment (yhirose) [inline peglib.h]: Nice found!

File: peglib.h [hunk 2]
```diff
@@ -1295,25 +1341,104 @@ class LiteralString : public Ope
...
             }
         }
     }
 
-    CharacterClass(const std::vector<std::pair<char32_t, char32_t>>& ranges) : ranges_(ranges) {}
+public:
+    CharacterClass(const std::string& str)
+    {
+        // first char in char class might negate char class
+        negated = str.front() == '^';
```
Comment (yhirose) [inline peglib.h]: 'negated' character class support seems to be common in many PEG libraries. But this should be recognized in the PEG grammar.

**Comment by yhirose** [comment]
@mumme74,  thank you for bringing up this matter. I actually didn't support such enhanced 'character class' notations, rather I was actually thinking to support the full Unicode regular expression in the 'literal' which seemed more useful to users instead of introducing 'non-official' PEG notations. https://github.com/yhirose/cpp-peglib#todo

I took a look at how other PEG parser libraries handle this matter, and found each library takes its own way since [the Bryan Ford's original PEG paper](https://bford.info/pub/lang/peg.pdf) doesn't provide such notations. Here are some information:
https://nim-lang.org/docs/pegs.html
https://relishapp.com/jarhart/rattler/docs/extended-matching-syntax/posix-character-classes
https://github.com/pointlander/peg#peg-file-syntax
https://github.com/pointlander/peg/issues/28
That's why I am hesitated to provide the enhanced character class functionality, yet.

As for the 'negated' character class in the pull request, I feel now it is OK to support since a number of PEG libraries support it with the same notation `[^a-z]`. But the syntax should be handled when parsing a PEG grammar.

I recognized  a bug with an orphaned backslash at the end of character class. That's should be fixed.

Thank you for all your efforts to make cpp-peglib more usable!

**Comment by yhirose** [comment]
2180657eead78f526defecc7958a86688d0d4041

**Comment by yhirose** [comment]
3dc0205ffa348840e022fa73b3596bcd45eb2a3d
2180657eead78f526defecc7958a86688d0d4041

File: peglib.h (hunk 1 above)
Comment (yhirose) [reply peglib.h]: I actually decided to report it as an error just like Regex.

**Comment by mumme74** [comment]
Glad you found some ideas inspirational!

I realize I don´t have enough knowledge to say anything about how PEG grammar might be extended. For me this was a way for me to learn and train myself during a christmas holiday.
I am not a trained programmer, just a self taught hobby person. Digging into a well architectured library written in modern c++ was an eye opening experience!

I´m currently deep involved in my daytime work. (teaching automotive related stuff)
Sorry I don't respond for extended times, been busy..

I trust you can make the best choices and steer cpp-peglib forward!
Just glad I could help with some inspiration.

Regards
Fredrik Johansson

**Comment by yhirose** [comment]
@mumme74, I really appreciated that you are interested in this library and gave me good ideas!

**Pull request closed without merging**